### PR TITLE
Drive Documentation

### LINKED ISSUE #10: Drive Subsystem
- [ ] List of Inputs (Sensors)
- [ ] List of Outputs (Motors, Pneumatics)
    - [ ] `RobotMap`
- [ ] List of Actions (Commands)
- [ ] Integrate Commands into Robot class

### PR DESCRIPTION
**Pull request by jeffaburke**
Closes #10

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 1]
```diff
@@ -10,19 +10,50 @@
...
+ * 
+ * 3) What interactions does it have with other subsystems? Does not have any
+ * interactions with other systems other than the space it shares on the robot
+ * 
+ * 4) How is it triggered -> OI? Triggered by motors to make the robots wheels
+ * spin a certain direction.
+ * 
+ * 5) Does it store any state? Right and Left motor groups.
+ * 
+ * 6) Sensors? Speed Controllers
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Drive.java]: These aren't sensors

Comment (msoucy) [reply src/main/java/frc/robot/subsystems/Drive.java]: Are there encoders? Gyros?

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 2]
```diff
@@ -10,19 +10,50 @@
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
 import frc.robot.maps.RobotMap;
 
+/**
+ * 1) What does it do? Makes motors turn a certain amount depending on how much
+ * an axis is triggered
+ * 
+ * 2) What modes does it have? Has a mode to differentiate between right and
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Drive.java]: This is not a mode - those are just members.

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 3]
```diff
@@ -10,19 +10,50 @@
...
+ * left motor groups.
+ * 
+ * 3) What interactions does it have with other subsystems? Does not have any
+ * interactions with other systems other than the space it shares on the robot
+ * 
+ * 4) How is it triggered -> OI? Triggered by motors to make the robots wheels
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Drive.java]: Triggered **by** motors?

**Review by msoucy — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 4]
```diff
@@ -2,33 +2,69 @@
...
+        rightMotorGroup = map.getRight();
+        leftMotorGroup = map.getLeft();
         driveTrain = new DifferentialDrive(leftMotorGroup, rightMotorGroup);
     }
 
+    /**
+     * Maps the drive axises
+     * 
+     * @param forward defines the forward direction
+     * @param turn    defines the direction to turn
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Drive.java:51]: These don't really "define" the directions, they provide the amount of forward/turning speed.

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 5]
```diff
@@ -2,33 +2,69 @@
...
     public CommandBase drive(DoubleSupplier forward, DoubleSupplier turn) {
         return new RunCommand(() -> {
             double yAxis = forward.getAsDouble();
             double xAxis = turn.getAsDouble();
-
             driveTrain.arcadeDrive(yAxis, xAxis);
         }, this);
     }
+
+    public CommandBase driveFlip(DoubleSupplier forward, DoubleSupplier turn) {
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Drive.java]: Maybe `reversedDrive`
This should have the same comment structure as the `drive` command. 

It should also be bound to a button with `toggleWhenActive`.

**Review by bot190 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/maps/TempestMap.java [hunk 6]
```diff
@@ -53,4 +46,10 @@ public LiftMap getLiftMap() {
         return new LiftMap() {
         };
     }
+
+    @Override
+    public IntakeMap getIntakeMap() {
```
Comment (bot190) [inline src/main/java/frc/robot/maps/TempestMap.java]: This shouldn't be here.

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 7]
```diff
@@ -2,33 +2,69 @@
...
+        return new RunCommand(() -> {
+            double yAxis = forward.getAsDouble();
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Drive.java]: This should really just be providing a new DoubleSupplier to `drive` that reverses the direction.

**Review by bot190 — CHANGES_REQUESTED** [review]
A few comments to cleanup the `driveFlip` command

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 8]
```diff
@@ -45,4 +45,4 @@ public StartEndCommand discharge() {
         }, this);
     }
 
-}
\ No newline at end of file
+    }
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Intake.java:48]: This change shouldn't be here.

File: src/main/java/frc/robot/subsystems/Drive.java [hunk 9]
```diff
@@ -2,32 +2,60 @@
...
+ */
 public class Drive extends SubsystemBase {
 
     private final SendableSpeedController rightMotorGroup;
     private final SendableSpeedController leftMotorGroup;
     private final DifferentialDrive driveTrain;
 
-    public Drive(RobotMap.DriveMap map) {
+    /**
+     * Defines the left and right motor(s) and then puts them into a differential
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Drive.java]: This doesn't `Define` motors, it gets them from the robot map. Please update this comment.

File: src/main/java/frc/robot/maps/RobotMap.java [hunk 10]
```diff
@@ -1,23 +1,15 @@
...
-
-        public SendableSpeedController right() {
-            return new MockSpeedController();
-        }
-    }
 
-    public DriveMap getDriveMap() {
-        return new DriveMap();
+    public DifferentialDriveMap getDriveMap() {
+        return null;
```
Comment (bot190) [inline src/main/java/frc/robot/maps/RobotMap.java]: This seems dangerous. We should provide a mock drive map, probably in ChopShopLib. For now this should provide mock speed controllers

**Review by bot190 — CHANGES_REQUESTED** [review]
A few more changes.

File: src/main/java/frc/robot/maps/HyperionMap.java [hunk 11]
```diff
@@ -1,26 +1,65 @@
...
 @RobotMapFor("Hyperion")
 public class HyperionMap extends RobotMap {
 
-    @Override
-    public DriveMap getDriveMap() {
-        return new DriveMap() {
+    public DifferentialDriveMap getDriveMap() {
+        return new DifferentialDriveMap() {
+
+<<<<<<< HEAD
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/HyperionMap.java]: This doesn't seem like it's in the right place...

File: src/main/java/frc/robot/maps/TempestMap.java [hunk 12]
```diff
@@ -1,39 +1,89 @@
 package frc.robot.maps;
 
+<<<<<<< HEAD
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/TempestMap.java]: Merge conflict

File: src/main/java/frc/robot/maps/TempestMap.java [hunk 13]
```diff
@@ -1,39 +1,89 @@
...
+import com.chopshop166.chopshoplib.maps.DifferentialDriveMap;
+import com.chopshop166.chopshoplib.outputs.EncodedSpeedController;
 import com.chopshop166.chopshoplib.outputs.SendableSpeedController;
+import com.chopshop166.chopshoplib.sensors.MockEncoder;
 import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
 
-import edu.wpi.first.wpilibj.SpeedControllerGroup;
-import edu.wpi.first.wpilibj.Talon;
-
+<<<<<<< HEAD
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/TempestMap.java]: Merge conflict

File: src/main/java/frc/robot/maps/TempestMap.java [hunk 14]
```diff
@@ -1,39 +1,89 @@
...
 @RobotMapFor("Tempest")
 public class TempestMap extends RobotMap {
+>>>>>>> Drive
 
     @Override
-    public DriveMap getDriveMap() {
-        return new DriveMap() {
+    public DifferentialDriveMap getDriveMap() {
+        return new DifferentialDriveMap() {
+<<<<<<< HEAD
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/TempestMap.java]: Merge conflict

File: src/main/java/frc/robot/maps/TempestMap.java [hunk 15]
```diff
@@ -1,39 +1,89 @@
...
-            public SendableSpeedController left() {
-                return SendableSpeedController.wrap(new SpeedControllerGroup(new WPI_TalonSRX(4), new WPI_TalonSRX(1)));
+            public EncodedSpeedController getLeft() {
+                final SendableSpeedController leftGroup = SendableSpeedController.group(new WPI_TalonSRX(1),
+                        new WPI_TalonSRX(4));
+                return EncodedSpeedController.join(leftGroup, new MockEncoder());
+>>>>>>> Drive
             }
 
+<<<<<<< HEAD
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/TempestMap.java]: Merge conflict

File: src/main/java/frc/robot/subsystems/Intake.java (hunk 8 above)
Comment (msoucy) [reply src/main/java/frc/robot/subsystems/Intake.java:48]: Please undo this

**Review by msoucy — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/maps/HyperionMap.java [hunk 16]
```diff
@@ -1,26 +1,32 @@
...
 
-import edu.wpi.first.wpilibj.SpeedControllerGroup;
-
 @RobotMapFor("Hyperion")
 public class HyperionMap extends RobotMap {
 
-    @Override
-    public DriveMap getDriveMap() {
-        return new DriveMap() {
+    public DifferentialDriveMap getDriveMap() {
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/HyperionMap.java:13]: This should be `@Override`

**Review by bot190 — APPROVED** [review]

File: src/main/java/frc/robot/maps/HyperionMap.java (hunk 16 above)
Comment (msoucy) [reply src/main/java/frc/robot/maps/HyperionMap.java:13]: Does this chassis have a gyro?

File: src/main/java/frc/robot/maps/TempestMap.java [hunk 17]
```diff
@@ -1,39 +1,32 @@
...
 public class TempestMap extends RobotMap {
 
     @Override
-    public DriveMap getDriveMap() {
-        return new DriveMap() {
-            @Override
-            public SendableSpeedController left() {
-                return SendableSpeedController.wrap(new SpeedControllerGroup(new WPI_TalonSRX(4), new WPI_TalonSRX(1)));
-            }
+    public DifferentialDriveMap getDriveMap() {
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/TempestMap.java:14]: Does this chassis have a gyro?

**Review by msoucy — CHANGES_REQUESTED** [review]

**Review by msoucy — APPROVED** [review]

**Pull request merged**